mqtt-streaming/src/main/scala/akka/stream/alpakka/mqtt/streaming/impl/ClientConnector.scala

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (C) 2016-2018 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.mqtt.streaming
+package impl
+
+import akka.actor.typed.{ActorRef, Behavior}
+import akka.actor.typed.scaladsl.Behaviors
+import akka.annotation.InternalApi
+import akka.util.ByteString
+
+/*
+ * A client connector is a Finite State Machine that manages MQTT client
+ * session state. A client connects to a server, subscribes/unsubscribes
+ * from topics to receive publications on and publishes to its own topics.
+ */
+@InternalApi private[streaming] object ClientConnector {
+
+  type ConnectData = Option[_]
+
+  /*
+   * Construct with the starting state
+   */
+  def apply(settings: MqttSessionSettings): Behavior[Event] =
+    disconnected(Uninitialized(settings))
+
+  // Our FSM data and events
+
+  sealed abstract class Data(val settings: MqttSessionSettings)
+  final case class Uninitialized(override val settings: MqttSessionSettings) extends Data(settings)
```

Comment:
The code linting is requiring that I have override val.

Reply:
Really? We should cachange that. @2m ?

Reply:
It's not me, it's scalac. :)

`override val` is required because that case class extends an abstract class. Changing `Data` to `sealed trait` fixes it:

```
sealed trait Data {
  val packetIdAllocator: ActorRef[PacketIdAllocator.Request]
  val settings: MqttSessionSettings
}
```

Reply:
Nice, did not know that. Thanks!

Reply:
@huntc is using a trait an option?

Reply:
Perhaps abstract classes vs traits no longer have much meaning in Scala given that Java 8 byte code permits traits to have fields. However, the purist in me still thinks there is a "isa" vs "hasa" argument to be had... and there's precedent for `abstract class` elsewhere in Alpakka. :-)

All of that aside, traits don't always save you code. Look at this before/after scenario in our code:

*before*
```scala
sealed abstract class ControlPacket(val packetType: ControlPacketType, val flags: ControlPacketFlags)

case object Reserved1 extends ControlPacket(ControlPacketType.Reserved1, ControlPacketFlags.ReservedGeneral)

case object Reserved2 extends ControlPacket(ControlPacketType.Reserved2, ControlPacketFlags.ReservedGeneral)
```

*after*
```scala
sealed trait ControlPacket {
  def packetType: ControlPacketType
  def flags: ControlPacketFlags
}

case object Reserved1 extends ControlPacket {
  val packetType: ControlPacketType = ControlPacketType.Reserved1
  val flags: ControlPacketFlags = ControlPacketFlags.ReservedGeneral
}

case object Reserved2 extends ControlPacket {
  val packetType: ControlPacketType = ControlPacketType.Reserved2
  val flags: ControlPacketFlags = ControlPacketFlags.ReservedGeneral
}
```

Reply:
I consider the second more readable, because long lines give me hard times. But anyway, that's just me.